src/plugins/alt-titles.ts

```diff
@@ -0,0 +1,217 @@
+      to: Math.max(
+        state.runningTitle[1] + state.runningTitle[0].nodeSize,
+        state.shortTitle[1] + state.shortTitle[0].nodeSize
+      ),
+    }
+    // if selection overlaps with the range of alt titles, the selection is in the titles
+    return Math.max(from, range.from) <= Math.min(to, range.to)
+  }
+  return false
+}
+
+export const altTitlesKey = new PluginKey<PluginState>('altTitles')
+
+export default () => {
+  return new Plugin<PluginState>({
+    key: altTitlesKey,
+    state: {
+      init: (_, state) => {
+        return { collapsed: true, ...getTitlesData(state.doc) }
+      },
+      apply: (tr, value) => {
+        let newState = value
+        if (tr.docChanged) {
+          newState = { ...newState, ...getTitlesData(tr.doc) }
+        }
+        if (
+          tr.selectionSet &&
+          selectionInAltTitles(tr.selection.from, tr.selection.to, newState)
+        ) {
+          // if selection was set in the titles - open them
+          newState = { ...newState, collapsed: false }
+        }
+        if (tr.getMeta(altTitlesKey)) {
+          newState = { ...newState, ...tr.getMeta(altTitlesKey) }
+        }
+        return newState
+      },
+    },
+    appendTransaction: (transactions, _, newState) => {
+      // in appendTransaction we check if alt_titles nodes exist before opening them for the first time because they are optional
+      const tr = newState.tr
+      if (
+        !transactions.some((tr) => tr.getMeta(altTitlesKey)) ||
+        !altTitlesKey.getState(newState)
+      ) {
+        return null
+      }
+
+      const { titleEnd, runningTitle, shortTitle } =
+        altTitlesKey.getState(newState)!
+      const schema = newState.schema
+
+      if (!titleEnd) {
+        return null
+      }
+      if (!runningTitle) {
+        const title = schema.nodes.alt_title.create({
+          type: 'running',
+        })
+        tr.insert(titleEnd, title)
+      }
+      if (!shortTitle) {
+        const title = schema.nodes.alt_title.create({ type: 'short' })
+        const newPos = tr.mapping.map(titleEnd)
+        tr.insert(newPos, title)
+      }
+
+      return tr
+    },
+    props: {
+      decorations: (state) => {
+        const decorations: Decoration[] = []
+        const pState = altTitlesKey.getState(state)
+        if (!pState) {
+          return DecorationSet.empty
+        }
+
+        if (!pState.collapsed) {
+          let lastTitleEnd: number | undefined
+          state.doc.descendants((node, pos) => {
+            if (node.type === state.schema.nodes.alt_title) {
+              decorations.push(
+                Decoration.node(pos, pos + node.nodeSize, {
+                  class: 'manuscript-alt-title-open',
+                })
+              )
+              lastTitleEnd = pos + node.nodeSize
+            }
+          })
+          if (lastTitleEnd) {
+            decorations.push(
+              Decoration.widget(
+                lastTitleEnd,
```

Comment:
Can't you depend on the `titleEnd`?

Reply:
titleEnd is the end of titleNode. 'lastTitleEnd' the position after the last alt-title node. We need to insert this widget exactly after the last alt-title node.